### PR TITLE
don't add extra shuffle in DataLoader2 if one is present

### PR DESCRIPTION
**Pull request by pmeier**
Without this, `DataLoader2` will just add an `Shuffler` to the end of the datapipe if `shuffle=True`:

```py
from torch.utils.data.dataloader_experimental import DataLoader2

from torchdata.datapipes.iter import IterableWrapper, IterDataPipe, Shuffler


class Sorter(IterDataPipe):
    def __init__(self, datapipe):
        self.datapipe = datapipe

    def __iter__(self):
        return iter(sorted(self.datapipe))


data = list(range(1000))
dp = IterableWrapper(data)
dp = Shuffler(dp).set_shuffle(False)
dp = Sorter(dp)

dl2 = DataLoader2(dp, shuffle=True, batch_size=None)

assert list(dl2) == data  # fails unless you hit a lucky random seed
```

This example is somewhat non-sensical, but demonstrates we cannot simply add a `Shuffler`.

### CODE REVIEW

File: test/test_dataloader.py [hunk 1]
```diff
@@ -2173,26 +2173,46 @@ def test_basics(self):
...
+
     def test_shuffle(self):
         items = list(range(1000))
         dp = IterableWrapper(items).sharding_filter().shuffle()
 
         dl = DataLoader2(dp, batch_size=None, num_workers=2, shuffle=False)
         self.assertEqual(items, list(dl))
 
-        dl = DataLoader(dp, batch_size=None, num_workers=2, shuffle=False,
+        dl = DataLoader2(dp, batch_size=None, num_workers=2, shuffle=False,
```
Comment (pmeier) [inline test/test_dataloader.py:2190]: I'm guessing this was a mistake given that this test case is specifically for `DataLoader2`?

File: torch/utils/data/dataloader_experimental.py [hunk 2]
```diff
@@ -87,8 +89,11 @@ def __new__(cls,
                 raise Exception(
                     'sampler is not yet supported by DataPipes')
             datapipe = dataset
-            if shuffle:
-                # Enforce at least one shuffle in the graph
+            # Enforce at least one shuffle in the graph if shuffle=True
+            if shuffle and not any(
+                    isinstance(dp, Shuffler)
+                    for dp in torch.utils.data.graph_settings.get_all_graph_pipes(traverse(dataset, only_datapipe=True))
+            ):
```
Comment (pmeier) [inline torch/utils/data/dataloader_experimental.py:96]: 1. We should probably warn if we need to add a `Shuffler`, since the default buffer size may not be sufficient for the task at hand. Otherwise the user might see significantly degraded performance without any indication.
2. It would merge this into 

    https://github.com/pytorch/pytorch/blob/fa1a41ca718fd45a29ac1655861f40703c875c1b/torch/utils/data/graph_settings.py#L30

    given that it is already looking for `Shuffler`'s. Not sure about BC though. Is this considered public?

Comment (ejguan) [reply torch/utils/data/dataloader_experimental.py:96]: 1. I totally agree
2. I think it's fine. This is not public and we even have a plan to bring DataLoader2 into TorchData and spin off TorchData from PyTorch

**Review by ejguan — APPROVED** [review]
It totally makes sense. Thanks.

**Comment by ejguan** [comment]
Hey @pmeier,
Can I ask for a follow up on this PR to change default value for shuffle of `DataLoader` to `None` rather than `False`. For the case of `IterDataPipe` and when `shuffle` is `None`, we won't override the shuffle setting for the datapipe graph,
